lib/events/complete.go

```diff
@@ -140,6 +143,10 @@ func (u *UploadCompleter) CheckUploads(ctx context.Context) error {
 		u.log.Debugf("Completed upload %v.", upload)
 		completed++
 		uploadData := u.cfg.Uploader.GetUploadMetadata(upload.SessionID)
+		err = u.emitSessionEndEvent(ctx, uploadData)
+		if err != nil {
+			return trace.Wrap(err)
+		}
```

Comment:
IIUC `session.end` is emitted before the stream completes, meaning that we could have a situation where something goes wrong between `session.end` emission, and completion.  In that case, this logic would cause a duplicate `session.end` to be added.  I think we need to be checking for a `session.end` and only adding one if none exists.  Looks like `emitSessionEndEvent` is loading the session events anyhow, so how about renaming it to something like `ensureSessionEndEvent`, and having perform the check internally?